### ISSUE TITLE
Questions about DARTS

### ISSUE THREAD

**Issue by promach** [D-X-Y/AutoDL-Projects#99]
1. For DARTS complexity analysis, anyone have any idea how to derive the *(k+1)\*k/2* expression ?  Why 2 input nodes ?  How will the calculated value change if graph isomorphism is considered ?  Why "2+3+4+5" learnable edges ?  If there is lack of connection, the paper should not add 1 which does not actually contribute to learnable edges configurations at all ?

2. Why need to train the weights for normal cells and reduction cells **separately** as shown in Figures 4 and 5 below ?

3. How to arrange the nodes such that the NAS search will actually converge with minimum error ?   Note: Not all nodes are connected to each and every other nodes

4. Why is [GDAS](https://arxiv.org/pdf/1910.04465.pdf#page=7) 10 times faster than DARTS ?

![DARTS_normal_reduction_cells](https://user-images.githubusercontent.com/3324659/116770930-e97d4d80-aa79-11eb-8aae-ecf867aabe49.png)

![DARTS_complexity_analysis](https://user-images.githubusercontent.com/3324659/116770928-e2eed600-aa79-11eb-90b9-f00546958b12.png)

**Comment by D-X-Y** [comment]
Thanks for pointing out these questions.

(1). (k+1)k/2 is because for the k-th node, you have (k+1) preceding nodes. Selecting two from them has C(K+1, 2) possibilities. 2 input nodes are pre-defined according to human expert's experience. If isomorphism is considered, you need another way to represent this DAG. Before pruning the fully-connected graph into "2-input-nodes version", each node has (k+1) preceding nodes and has (k+1) edges -> (1+1) + (2+1) + (3+1) + (4+1) = 14 learable edges.

(2). We hypothesis the normal cell and reduction cell will have a very different topology structure

**Comment by promach** [comment]
Why **for the k-th node, you have (k+1) preceding nodes.** ?

**Comment by D-X-Y** [comment]
Because for each cell, they also allow the output of two previous cells as inputs, so for the 1-th first node in a cell, its preceding nodes are `[last-cell-outputs, second-last-cell-outputs]`. For the second node, it is: `[last-cell-outputs, second-last-cell-outputs, first-node-outputs]`

**Comment by promach** [comment]
For second node, what is the difference between `last-cell-outputs` and `first-node-outputs` ?  

![image](https://user-images.githubusercontent.com/3324659/117573177-20ea9a80-b109-11eb-9418-16e22e684164.png)

**Comment by D-X-Y** [comment]
The `last-cell-outputs` is the output of green box `c_{k-1}`.
The `first-node-outputs` is the output of blue box `0`.

**Comment by promach** [comment]
> if you add gumbel distributed noise to logits and take the argmax, the gumbel noise is the exact right distribution that it is the same as softmaxing the logits and sampling from the discrete distribution defined by those probabilities

Someone told me that the above, but I am not familiar with gumbel and how it actually helps to speed up GDAS with respect to DARTS.  I suppose it is the gumbel-max trick mentioned in the paper.  I do not quite understand expressions (3) and (5) in the GDAS paper.

**Comment by promach** [comment]
@D-X-Y Could you comment on this reply on your [Gumbel-Max code implementation](https://discuss.pytorch.org/t/gumbel-max-implementation/121320/2?u=promach) ?

@Unity05 was suggesting to use [softargmax](https://bouthilx.wordpress.com/2013/04/21/a-soft-argmax/)

**Comment by Unity05** [comment]
Hi,
I was just explaining that the temperature you're using uses the same basic idea as softargmax.

**Comment by promach** [comment]
@D-X-Y 

in your [coding](https://github.com/D-X-Y/AutoDL-Projects/blob/main/lib/models/cell_searchs/search_model_gdas.py#L121), would you be able to describe how the logic of `hardwts = one_h - probs.detach() + probs` is used in the forward search function `feature = cell.forward_gdas(feature, hardwts, index)` ?  

I mean the computation **logic** for `hardwts` is a bit weird or strange.  Why `hardwts` need to make use of both `one_h` and `probs` ?

**Comment by promach** [comment]
For the question on `hardwts` , see the note section inside https://pytorch.org/docs/stable/nn.functional.html#gumbel-softmax

```
The main trick for hard is to do y_hard - y_soft.detach() + y_soft

It achieves two things: 
- makes the output value exactly one-hot (since we add then subtract y_soft value) 
- makes the gradient equal to y_soft gradient (since we strip all other gradients)
```

@D-X-Y  by the way, would DARTS or GDAS be able to eliminate the need for reduction cells since [PNASNet](https://arxiv.org/pdf/1712.00559.pdf#page=5) only learns normal cells ?

![image](https://user-images.githubusercontent.com/3324659/118696075-7c6f0380-b840-11eb-84b5-cb3760139bd4.png)

**Comment by promach** [comment]
Solution:

```
we do not distinguish between Normal and Reduction cells, 
but instead emulate a Reduction cell by using a Normal cell with stride 2
```
So, in this case, I suppose I could use only single type of weights for both normal cells and reduction cells ?

As for algorithm 1, how is A different from W ?
Note: The corresponding notation meaning explanation after equations (3) and (4) of the [paper](https://arxiv.org/pdf/1910.04465.pdf#page=4) is very confusing to me.

![image](https://user-images.githubusercontent.com/3324659/119068355-9c582000-ba16-11eb-94e5-fb43a56392ee.png)

**Comment by D-X-Y** [comment]
> @D-X-Y
> 
> in your [coding](https://github.com/D-X-Y/AutoDL-Projects/blob/main/xautodl/models/cell_searchs/search_model_gdas.py#L121), would you be able to describe how the logic of `hardwts = one_h - probs.detach() + probs` is used in the forward search function `feature = cell.forward_gdas(feature, hardwts, index)` ?
> 
> I mean the computation **logic** for `hardwts` is a bit weird or strange.
> Why `hardwts` need to make use of both `one_h` and `probs` ?
> Why one of the `probs` need `detach()` ?
> 
> Besides, why would gumbel-max computation need a `while` loop ? I suppose you are using Gumbel(0, 1) ?
> 
> How exactly gumbel-max [transforms](https://arxiv.org/pdf/1910.04465.pdf#page=4) equation (3) into equation (5) ?

Sorry for the late reply, I'm a little bit busy these days.

`hardwts = one_h - probs.detach() + probs` aims to make `hardwts` has the same gradients as `probs` yet still keeps the one-hot values.
The `while` loop is a trick added by myself, which is to avoid very rare cases of `NAN`

**Comment by D-X-Y** [comment]
> For the question on `hardwts` , see the note section inside https://pytorch.org/docs/stable/nn.functional.html#gumbel-softmax
> 
> ```
> The main trick for hard is to do y_hard - y_soft.detach() + y_soft
> 
> It achieves two things: 
> - makes the output value exactly one-hot (since we add then subtract y_soft value) 
> - makes the gradient equal to y_soft gradient (since we strip all other gradients)
> ```
> 
> @D-X-Y by the way, why [PNASNet](https://arxiv.org/pdf/1712.00559.pdf#page=5) mention `Note that we learn a single cell type instead of distinguishing between Normal and Reduction cell.` ?
> 
> ![image](https://user-images.githubusercontent.com/3324659/118696075-7c6f0380-b840-11eb-84b5-cb3760139bd4.png)

Yes, I borrow the idea of how to implement gumbel from PyTorch with a few modifications.

For `PNAS`, you may need to email their authors for the detailed reasons.

**Comment by D-X-Y** [comment]
> Solution:
> 
> ```
> we do not distinguish between Normal and Reduction cells, 
> but instead emulate a Reduction cell by using a Normal cell with stride 2
> ```
> 
> So, in this case, I suppose I could use only single type of weights for both normal cells and reduction cells ?
> 
> As for algorithm 1, how is A different from W ?
> Note: The corresponding notation meaning explanation after equations (3) and (4) of the [paper](https://arxiv.org/pdf/1910.04465.pdf#page=4) is very confusing to me.
> 
> ![image](https://user-images.githubusercontent.com/3324659/119068355-9c582000-ba16-11eb-94e5-fb43a56392ee.png)

Yes, in this case, the architecture weights for normal cells and reduction cells are shared.
`A` is the architecture weights -- the logits assigned for each candidate operation.
`W` is the weights of the supernet -- the weights for convolution layers, etc.

**Comment by promach** [comment]
@D-X-Y I am bit confused with the [difference](https://github.com/D-X-Y/AutoDL-Projects/issues/99#issuecomment-835713058) between `cell` and `node`

**Comment by D-X-Y** [comment]
> @D-X-Y I am bit confused with the [difference](https://github.com/D-X-Y/AutoDL-Projects/issues/99#issuecomment-835713058) between `cell` and `node`
> 
> Edit: I think I got it now. A single cell contains 4 distinct nodes
> 
> By the way, in Algorithm 1, why GDAS updates `W` before `A` ?

I feel it does not matter? Updating `W, A, W, A, W, A` or `A, W, A, W, A, W` would not make a big difference?

**Comment by promach** [comment]
For GDAS, would https://networkx.org/documentation/stable/tutorial.html#multigraphs be suitable for both forward inference and backward propagation ?

**Comment by D-X-Y** [comment]
I'm not familiar with `networkx` and can not comment on that.

**Comment by promach** [comment]
@D-X-Y I am confused as in how https://github.com/D-X-Y/AutoDL-Projects/blob/main/xautodl/models/cell_searchs/search_model_gdas.py implemented multiple parallel connections between nodes

**Comment by promach** [comment]
@D-X-Y I am confused as in how equation (7) is an approximation of equation (5) as described in [gdas paper](https://arxiv.org/pdf/1910.04465.pdf#page=4) ?

**Comment by D-X-Y** [comment]
@promach 
The difference between $h$ in Eq.(5) and Eq.(7) is that:
- In Eq.(5), it is a one-hot vector.
- In Eq.(7), it is a soft-probability vector with Gumbel noise.

**Comment by promach** [comment]
@D-X-Y  in normal backpropagation, there is only a single edge in between two nodes.

However in [GDAS](https://github.com/D-X-Y/AutoDL-Projects/issues/99#issuecomment-835802887), there are multiple parallel edges in between two nodes.

So, how to perform backpropagation for GDAS or more generally, Network Architecture Search (NAS) ?

![image](https://user-images.githubusercontent.com/3324659/120290637-1f218a80-c2f5-11eb-95d2-265dde523759.png)

**Comment by promach** [comment]
For https://github.com/D-X-Y/AutoDL-Projects/issues/99#issuecomment-845789377 ,  how do you actually update both `W` and `A` simultaneously in a single epoch ?  

Could you point me to the relevant code for the *update* portion ?  
Did you only use a single `def forward()` function for both `W` and `A` ?

![GDAS algorithm](https://user-images.githubusercontent.com/3324659/119068355-9c582000-ba16-11eb-94e5-fb43a56392ee.png)

**Comment by D-X-Y** [comment]
@promach , at a single iteration, we will first update `W` and then update `A`.
Please see the codes here: https://github.com/D-X-Y/AutoDL-Projects/blob/main/exps/NAS-Bench-201-algos/GDAS.py#L49

**Comment by promach** [comment]
If update `W` first, then only update `A` ,  the question is should I train the convolution kernel weights `W` based on the trained **best** edges for `A` ?

**Comment by D-X-Y** [comment]
I feel it does not matter for the order of `W` and `A`. As if you look at multiple iterations, it will be `W` -> `A` -> `W` -> `A` ->`W` -> `A` -> `W` -> `A` -> `W` -> `A` -> `W` -> `A` .... Whether the first one is `W` or `A` would not make a big difference.

**Comment by promach** [comment]
the issue lingering in my head is that if `W` is to optimized FIRST, should `W` be trained under which exact `A` result ?

**Comment by D-X-Y** [comment]
What do you mean by `exact A`?

**Comment by promach** [comment]
If `W` is trained FIRST, which`A` should `W` training process uses as `W`'s architecture ?

**Comment by D-X-Y** [comment]
`A` is a set of variables indicating the architecture encoding. There only one `A` and no other options?

**Comment by D-X-Y** [comment]
You could have a look at the codes [here](https://github.com/D-X-Y/AutoDL-Projects/blob/main/exps/NAS-Bench-201-algos/GDAS.py#L49) and Would you mind clarifying what do you think the codes should be?

**Comment by promach** [comment]
Let me rephrase my question, how do you define `base_input` and `arch_input` ?

It seems to be different from how DARTS paper originally proposed.  See equations (5) and  (6) of [DARTS paper](https://arxiv.org/pdf/1806.09055.pdf#page=4)

![image](https://user-images.githubusercontent.com/3324659/121475409-f33e8d00-c9f7-11eb-9bf2-0b156cf6ec54.png)

**Comment by D-X-Y** [comment]
`base_inputs` are a batch of samples from the training data, `arch_inputs` are a batch of samples from the validation data.

Yes, following the DARTS paper, I should switch the order of updating `W` and `A`.

**Comment by promach** [comment]
during training for W, should I use a particular found architecture inside that particular epoch ? 
OR should I use the whole supernet ?

**Comment by D-X-Y** [comment]
It depends on the NAS algorithm. For DARTS, they use the whole supernet. For GDAS, we use an architecture candidate randomly sampled based on `A`.

**Comment by promach** [comment]
> For GDAS, we use an architecture candidate randomly sampled based on A.

The candidate is chosen using gumbel-argmax [(equation (5) and (6) of GDAS paper)](https://arxiv.org/pdf/1910.04465.pdf#page=4) , instead of chosen randomly.
Please correct me if wrong.

**Comment by D-X-Y** [comment]
`gumbel-argmax` is a kind of random? because the $o_{k}$ is randomly sampled from Gumbel(0, 1).

**Comment by promach** [comment]
For https://github.com/D-X-Y/AutoDL-Projects/issues/99#issuecomment-835802887 , there are two types of outputs from the blue node.

One type of (multiple edges) output connects to the input of the other blue nodes ?

Another type of (single edge) output connects directly to the yellow node ?

**Comment by promach** [comment]
It seems that both [ENAS](https://arxiv.org/pdf/1802.03268.pdf#page=8) and [PNAS](https://arxiv.org/pdf/1712.00559.pdf#page=4) just perform `add` and `concat` operations for the connection to the output node

![image](https://user-images.githubusercontent.com/3324659/123533252-5a717680-d746-11eb-930d-54eec0631232.png)


![image](https://user-images.githubusercontent.com/3324659/123533248-52193b80-d746-11eb-865d-09ef9c826dfc.png)

**Comment by promach** [comment]
@D-X-Y I implemented a draft code on GDAS,

However, could you advise whether [this edge weight training epoch mechanism](https://gist.github.com/promach/b6f526c56e20f029d68e6f9041c3f5c0/72c4b14b5e568beb72e413a69edfc3b7fd518f4b#file-gdas-py-L276-L280) will actually work for GDAS ?

**Comment by D-X-Y** [comment]
> For [#99 (comment)](https://github.com/D-X-Y/AutoDL-Projects/issues/99#issuecomment-835802887) , there are two types of outputs from the blue node.
> 
> One type of (multiple edges) output connects to the input of the other blue nodes ?
> 
> Another type of (single edge) output connects directly to the yellow node ?

Yes, you are right~

**Comment by D-X-Y** [comment]
> It seems that both [ENAS](https://arxiv.org/pdf/1802.03268.pdf#page=8) and [PNAS](https://arxiv.org/pdf/1712.00559.pdf#page=4) just perform `add` and `concat` operations for the connection to the output node
> 
> ![image](https://user-images.githubusercontent.com/3324659/123533252-5a717680-d746-11eb-930d-54eec0631232.png)
> 
> ![image](https://user-images.githubusercontent.com/3324659/123533248-52193b80-d746-11eb-865d-09ef9c826dfc.png)


Yes. DARTS also uses `add` for the intermediate nodes and `concat` for the final output node (https://github.com/D-X-Y/AutoDL-Projects/blob/main/xautodl/models/cell_searchs/search_cells.py#L251).

**Comment by D-X-Y** [comment]
> @D-X-Y I implemented a draft code on GDAS,
> 
> However, could you advise whether [this edge weight training epoch mechanism](https://gist.github.com/promach/b6f526c56e20f029d68e6f9041c3f5c0/72c4b14b5e568beb72e413a69edfc3b7fd518f4b#file-gdas-py-L276-L280) will actually work for GDAS ?

I personally feel the implementations are incorrect. I havn't fully checked the codes, but at least, the input for every cell/node should not be the same `forward_edge(train_inputs)`.

**Comment by promach** [comment]
How to code the forward pass function correctly for edge weight training ?

```
    # self-defined initial NAS architecture, for supernet architecture edge weight training
    def forward_edge(self, x):
        self.__freeze_f()
        self.__unfreeeze_w()

        return self.weights
```

**Comment by D-X-Y** [comment]
why do we return self.weights? Instead of return the value of using `weights` on `x`?
The logics of freeze and unfreeze are correct, but I do not understand `return ..`.

**Comment by promach** [comment]
I am not sure how to train edge weights, hence the question about `def forward_edge()`

Besides, I also suspect the forward pass function for architecture weight (step 1 inside DARTS Algorithm 1) might be incorrect as well because it only trains the neural network function's **internal** weight parameters instead of architecture weight.

Note: `self.f(x)` is something like `nn.Linear()` , `nn.Conv2d`

```
    # for NN functions internal weights training
    def forward_f(self, x):
        self.__unfreeze_f()
        self.__freeeze_w()

        # inheritance in python classes and SOLID principles
        # https://en.wikipedia.org/wiki/SOLID
        # https://blog.cleancoder.com/uncle-bob/2020/10/18/Solid-Relevance.html
        return self.f(x)
```

**Comment by promach** [comment]
Sorry, I misinterpreted the purpose of the two forward pass functions.

`forward_edge()` is for architecture weights (step 1), while `forward_f()` is for NN function's internal weights (step 2).

However, I am still not sure how to code for `def forward_edge(self, x)`

**Comment by promach** [comment]
@D-X-Y For ordinary NN training operation, we have some feature maps outputs.

However for the edge weights (NAS) training operation, there are no feature maps outputs though.  So, what should be fed into `x` for `forward_edge(x)` ?

**Comment by promach** [comment]
Is using `nn.Linear()` to [train edge weights](https://gist.github.com/promach/b6f526c56e20f029d68e6f9041c3f5c0#file-gdas-py-L92-L97) feasible for GDAS on a small GPU ?

```
    # self-defined initial NAS architecture, for supernet architecture edge weight training
    def forward_edge(self, x):
        self.__freeze_f()
        self.__unfreeeze_w()

        return self.linear(x)
```

**Comment by promach** [comment]
@D-X-Y I managed to get [my own GDAS code implementation](https://github.com/promach/gdas) up and running.

However, the loss stay the same which indicates the training process is still incorrect.

Could you advise ?

![image](https://user-images.githubusercontent.com/3324659/127015381-4960d094-8597-4a40-a007-223aea93055c.png)

**Comment by promach** [comment]
@D-X-Y looking at output of [graph.named_parameters()](https://paste.ubuntu.com/p/3SgTP9Q6P6/) , some of the [internal connections](https://github.com/promach/gdas/blob/main/gdas.py#L374-L438) within the super-net architecture are still not connected properly.  Any comments ?
